cmd/restic/cmd_ls.go

```diff
@@ -59,20 +74,89 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
+	// extract any specific directories to walk
+	var dirs []string
+	if len(args) > 1 {
+		dirs = args[1:]
+		for _, dir := range dirs {
+			if !strings.HasPrefix(dir, "/") {
+				return errors.Fatal("All path filters must be absolute, starting with a forward slash '/'")
+			}
+		}
+	}
+
 	ctx, cancel := context.WithCancel(gopts.ctx)
 	defer cancel()
-	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
-		Verbosef("snapshot %s of %v at %s):\n", sn.ID().Str(), sn.Paths, sn.Time)
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args[:1]) {
+		Verbosef("snapshot %s of %v filtered by %v at %s):\n", sn.ID().Str(), sn.Paths, dirs, sn.Time)
 
 		err := walker.Walk(ctx, repo, *sn.Tree, nil, func(nodepath string, node *restic.Node, err error) (bool, error) {
 			if err != nil {
 				return false, err
 			}
-
 			if node == nil {
 				return false, nil
 			}
+
+			// apply any directory filters
+			if len(dirs) > 0 {
+				nodeDir := path.Dir(nodepath)
+
+				// this first iteration ensures we do not traverse branches that
+				// are not in matching trees or will not lead us to matching trees
+				var walk bool
+				for _, dir := range dirs {
+					approachingMatchingTree := fs.HasPathPrefix(nodeDir, dir)
```

Comment:
This needs to use `nodepath` instead of `nodeDir`, otherwise one level is walked too much. 

When the nodepath is `/test/foo` and the filter dir is `/test/bar`, then this would still evaluate:
```go
fs.HasPathPrefix("/test", "/test/bar")
```

It should do the following:
```go
fs.HasPathPrefix("/test/foo", "/test/bar")
```